cmd/kubeadm/app/util/runtime_test.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	kubeadmapiv1alpha2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	"k8s.io/utils/exec"
+	fakeexec "k8s.io/utils/exec/testing"
+)
+
+func TestNewContainerRuntime(t *testing.T) {
+	execLookPathOK := fakeexec.FakeExec{
+		LookPathFunc: func(cmd string) (string, error) { return "/usr/bin/crictl", nil },
+	}
+	execLookPathErr := fakeexec.FakeExec{
+		LookPathFunc: func(cmd string) (string, error) { return "", fmt.Errorf("%s not found", cmd) },
+	}
+	cases := []struct {
+		execer    fakeexec.FakeExec
+		criSocket string
+		isDocker  bool
+		isError   bool
+	}{
+		{execLookPathOK, kubeadmapiv1alpha2.DefaultCRISocket, true, false},
+		{execLookPathOK, "unix:///var/run/crio/crio.sock", false, false},
+		{execLookPathErr, "unix:///var/run/crio/crio.sock", false, true},
+	}
+
+	for _, tc := range cases {
+		runtime, err := NewContainerRuntime(&tc.execer, tc.criSocket)
+		if err != nil {
+			if !tc.isError {
+				t.Errorf("unexpected NewContainerRuntime error. criSocket: %s, error: %v", tc.criSocket, err)
+			}
+			continue // expected error occurs, impossible to test runtime further
+		}
+		if tc.isError && err == nil {
+			t.Errorf("unexpected NewContainerRuntime success. criSocket: %s", tc.criSocket)
+		}
+		isDocker := runtime.IsDocker()
+		if tc.isDocker != isDocker {
+			t.Errorf("unexpected isDocker() result %v for the criSocket %s", isDocker, tc.criSocket)
+		}
+	}
+}
+
+func genFakeActions(fcmd *fakeexec.FakeCmd, num int) []fakeexec.FakeCommandAction {
+	var actions []fakeexec.FakeCommandAction
+	for i := 0; i < num; i++ {
```

Comment:
i think you can refactor this function to not use `i` and `num` and instead:
```
for _, action := range fcmd.RunScript {
    ...
}
```

Reply:
I don't think it's possible as there is at least one case where fcmd.CombinedOutputScript is passed to this function.

Reply:
can you point out the case where `len()` is needed, please?

Reply:
sure:
```
$ grep genFakeActions runtime_test.go 
func genFakeActions(fcmd *fakeexec.FakeCmd, num int) []fakeexec.FakeCommandAction {
		CommandScript: genFakeActions(&fcmd, len(fcmd.RunScript)),
		CommandScript: genFakeActions(&fcmd, len(fcmd.CombinedOutputScript)),
		CommandScript: genFakeActions(&fcmd, len(fcmd.RunScript)),
		CommandScript: genFakeActions(&fcmd, len(fcmd.RunScript)),
		CommandScript: genFakeActions(&fcmd, len(fcmd.RunScript)),
```
In most cases your suggestion to use fcmd.RunScript would work, but it wouldn't work when generating actions based on fcmd.CombinedOutputScript. Am I missing something?

Reply:
given other fields are used like `fcmd.CombinedOutputScript` to fetch a length, then `genFakeActions()` is better, as is.
agreed.